pkg/server/drain.go

```diff
@@ -139,79 +128,54 @@ func (s *adminServer) Drain(req *serverpb.DrainRequest, stream serverpb.Admin_Dr
+	// after the range holding the liveness record is marked as draining
+	// already). To prevent this, we need to briefly drop draining the
+	// stores so that KV ops can go through.
+	if err := s.stopDrain(ctx); err != nil {
+		return err
+	}
+	// Finally, perform the drain proper.
+	return s.drainNode(ctx)
 }
 
-// Undrain idempotently deactivates the given DrainModes on the Server in the
-// order in which they are supplied.
-// On success, returns any remaining active drain modes.
-func (s *Server) Undrain(
-	ctx context.Context, off []serverpb.DrainMode,
-) ([]serverpb.DrainMode, error) {
-	return s.doDrain(ctx, off, false /* setTo */)
+// isDraining returns true if either clients are being drained
+// or one of the stores on the node is not accepting replicas.
+func (s *Server) isDraining() bool {
+	return s.pgServer.IsDraining() || s.node.IsDraining()
 }
 
-func (s *Server) doDrain(
-	ctx context.Context, modes []serverpb.DrainMode, setTo bool,
-) ([]serverpb.DrainMode, error) {
-	for _, mode := range modes {
-		switch mode {
-		case serverpb.DrainMode_CLIENT:
-			if err := s.drainClients(ctx, setTo); err != nil {
-				return nil, err
-			}
-
-		case serverpb.DrainMode_LEASES:
-			if err := s.drainRangeLeases(ctx, setTo); err != nil {
-				return nil, err
-			}
-		default:
-			return nil, errors.Errorf("unknown drain mode: %s", mode)
-		}
-	}
-	var nowOn []serverpb.DrainMode
-	if s.pgServer.IsDraining() {
-		nowOn = append(nowOn, serverpb.DrainMode_CLIENT)
-	}
-	if s.node.IsDraining() {
-		nowOn = append(nowOn, serverpb.DrainMode_LEASES)
-	}
-	return nowOn, nil
-}
+// drainClients starts draining the SQL layer.
+func (s *Server) drainClients(ctx context.Context) error {
+	// Mark the server as draining in a way that probes to
+	// /health?ready=1 will notice.
+	s.grpc.setMode(modeDraining)
+	// Wait for drainUnreadyWait. This will fail load balancer checks and
+	// delay draining so that client traffic can move off this node.
+	time.Sleep(drainWait.Get(&s.st.SV))
 
-// drainClients starts draining the SQL layer if setTo is true, and
-// stops draining SQL if setTo is false.
-func (s *Server) drainClients(ctx context.Context, setTo bool) error {
-	if setTo {
-		s.grpc.setMode(modeDraining)
-		// Wait for drainUnreadyWait. This will fail load balancer checks and
-		// delay draining so that client traffic can move off this node.
-		time.Sleep(drainWait.Get(&s.st.SV))
-	} else {
-		// We're disableing the drain. Make the server
-		// operational again at the end.
-		defer func() { s.grpc.setMode(modeOperational) }()
-	}
 	// Since enabling the SQL table lease manager's draining mode will
 	// prevent the acquisition of new leases, the switch must be made
 	// after the pgServer has given sessions a chance to finish ongoing
 	// work.
-	defer s.leaseMgr.SetDraining(setTo)
-
-	if !setTo {
-		// Re-enable client connections.
-		s.distSQLServer.Undrain(ctx)
-		s.pgServer.Undrain()
-		return nil
-	}
+	defer s.leaseMgr.SetDraining(true /* drain */)
```

Comment:
Interesting choice of the defer (I know, not your choice, but nonetheless). Any idea why this wasn't just moved before the `return nil`? Doubt we want to do this if `pgServer.Drain()` errors.